Qt5

johnhaddon opened the pull request:
This makes Gaffer compatible with both Qt 4 and Qt 5, using [Qt.py](https://github.com/mottosso/Qt.py) to deal with some of the compatibility issues. I haven't updated the dependencies packages to include that yet, so the Travis build will fail here at first, but I wanted to get this PR up now so we can start testing at IE (I'll update the internal ticket with IE-specific details).

johnhaddon commented:
@ivanimanishi and I discussed this and decided it'd be good to get it merged, so we can do further testing as we do our day-to-day work on top of master. @donboie, @mattigruener, @danieldresser, could one of you give it a look over?

ivanimanishi commented on contrib/scripts/qt5.py:
```diff
@@ -0,0 +1,66 @@
+
+	with open( fileName ) as f :
+		text = "".join( f.readlines() )
+		#print text
```
The PR seems good to me.
My only comments are about this line (which is a commented out print that probably could be removed), and the documentation for this script.

I assume that the purpose of the script is to "Substitute QtWidgets for QtGui where needed" and "Replace deprecated `_qtImport` calls with `from Qt import` calls", which are comments found inside the code itself.

But I think it would be easier to get an idea of what the script is about if such comments were added at the top of the file as well.

johnhaddon replied:
Thanks Ivan - I've addressed this in 26b572c, using argparse to add proper command line help.

danieldresser-ie commented:
I haven't reviewed thoroughly, and I'm not too expert in Qt, but it all looks fine to me.

johnhaddon commented:
Yeah, the code changes are actually pretty simple - I think the important thing is getting as much testing done as we can. Travis isn't playing ball again though - Qt.py gives an error when its imported on OSX on Travis that I just don't get locally. Currently debugging very slowly by sending cut down jobs to Travis and waiting for them to pick up, see what happens, and then send another one...

danieldresser-ie commented:
I've been running with this today.

Issues so far:

* python/GafferUI/SplineWidget.py : Line 148 needs this fix:
```
-               brush.setMatrix( QtGui.QMatrix( float( rect.width() ) / numStops, 0, 0, 1, 0, 0 ) )
+               brush.setTransform( QtGui.QTransform( float( rect.width() ) / numStops, 0, 0, 1, 0, 0 ) )
```
( QMatrix is deprecated and not supported by Qt.py )

* Doc build at IE fails like this ( probably just some sort of env misconfiguration? ):
```
Running doc/source/Appendices/License/generate.py
Traceback (most recent call last):
  File "generate.py", line 2, in <module>
    import GafferUI
  File "/home/danield/apps/gaffer/0.33.3.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/__init__.py", line 121, in <module>
    from Enums import *
  File "/home/danield/apps/gaffer/0.33.3.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/Enums.py", line 40, in <module>
    from Qt import QtCore
ImportError: No module named Qt
```
( Probably not critical, I only ran it by accident because of the thing with the Gaffer SConstruct where "build" means "install the part you want" ).

* Jabuka needs updating
Probably not unexpected to you, but Ivan had hoped that because this was supposed to be backwards compatible, he wouldn't have to make major changes on his end.  I guess now that we're using Qt.py we would need to do a bunch of renames on Jabuka?  Ivan is concerned that with some of the fancy UI things Jabuka does, a simple rename might not be enough, and he doesn't really want to do this twice on the maintenance branch of Jabuka - but it will be a while before we're ready to run the new branch of Jabuka in production.  Hmm.

johnhaddon commented:
> I've been running with this today.

Thanks!

> * python/GafferUI/SplineWidget.py : Line 148 needs this fix:

Fixed in 73bd11f. I also added a contrib/scripts/qt5Check.py script that attempts to find similar problems rather than leaving them to runtime testing. This showed up one more which I fixed in 09e2017.

> * Doc build at IE fails like this ( probably just some sort of env misconfiguration? ):

Fixed in f402042.

> * Jabuka needs updating
> Probably not unexpected to you, but Ivan had hoped that because this was supposed to be backwards compatible, he wouldn't have to make major changes on his end. 

My intention was that Jabuka would run unchanged provided it was targeting Qt4 still. But I botched it. 
Fixed in c7ca934, and retested standalone and in Maya 2016 at IE.

johnhaddon commented:
Managed to reproduce the Travis OSX failure locally on my old Mac (still don't know why it works on my current one), and pushed one final fix which I think gets us to a mergeable state.
